scipy/_lib/_array_api.py

```diff
@@ -0,0 +1,239 @@
+"""
+import math
+import os
+
+import numpy as np
+# probably want to vendor it (submodule)
+import array_api_compat
+from array_api_compat import size
+import array_api_compat.numpy
+
+__all__ = ['array_namespace', 'as_xparray', 'as_xparray_namespace', 'isdtype']
+
+
+# SCIPY_ARRAY_API, array_api_dispatch is used by sklearn
+array_api_dispatch = os.environ.get("array_api_dispatch", False)
+SCIPY_ARRAY_API = os.environ.get("SCIPY_ARRAY_API", array_api_dispatch)
+SCIPY_DEVICE = os.environ.get("SCIPY_DEVICE", "cpu")
+
+_GLOBAL_CONFIG = {
+    "SCIPY_ARRAY_API": SCIPY_ARRAY_API,
+    "SCIPY_DEVICE": SCIPY_DEVICE,
+}
+
+
+def compliance_scipy(*arrays):
+    """Raise exceptions on known-bad subclasses.
+
+    The following subclasses are not supported and raise and error:
+    - `np.ma.MaskedArray`
+    - `numpy.matrix`
+    - Any array-like which is not Array API compatible
+    """
+    for array in arrays:
+        if isinstance(array, np.ma.MaskedArray):
+            raise TypeError("'numpy.ma.MaskedArray' are not supported")
+        elif isinstance(array, np.matrix):
+            raise TypeError("'numpy.matrix' are not supported")
+        elif not array_api_compat.is_array_api_obj(array):
+            raise TypeError("Only support Array API compatible arrays")
+        elif array.dtype is np.dtype('O'):
+            raise TypeError('object arrays are not supported')
+
+
+def _check_finite(array, xp):
+    """Check for NaNs or Infs."""
+    msg = "array must not contain infs or NaNs"
+    try:
+        if not xp.all(xp.isfinite(array)):
+            raise ValueError(msg)
+    except TypeError:
+        raise ValueError(msg)
+
+
+def array_namespace(*arrays):
+    """Get the array API compatible namespace for the arrays xs.
+
+    Parameters
+    ----------
+    *arrays : sequence of array_like
+        Arrays used to infer the common namespace.
+
+    Returns
+    -------
+    namespace : module
+        Common namespace.
+
+    Notes
+    -----
+    Thin wrapper around `array_api_compat.array_namespace`.
+
+    1. Check for the global switch: SCIPY_ARRAY_API. This can also be accessed
+       dynamically through ``_GLOBAL_CONFIG['SCIPY_ARRAY_API']``.
+    2. `compliance_scipy` raise exceptions on known-bad subclasses. See
+       it's definition for more details.
+
+    When the global switch is False, it defaults to the `numpy` namespace.
+    In that case, there is no compliance check. This is a convenience to
+    ease the adoption. Otherwise, arrays must comply with the new rules.
+    """
+    if not _GLOBAL_CONFIG["SCIPY_ARRAY_API"]:
+        # here we could wrap the namespace if needed
+        return array_api_compat.numpy
+
+    arrays = [array for array in arrays if array is not None]
+
+    compliance_scipy(*arrays)
+
+    return array_api_compat.array_namespace(*arrays)
+
+
+def as_xparray(
+    array, dtype=None, order=None, copy=None, *, xp=None, check_finite=True
+):
+    """Drop-in replacement for `np.asarray`.
```

Comment:
Where does the name `as_xparray` come from? I couldn't find it in https://github.com/scipy/scipy/issues/18286 nor on the array-api [docs](https://data-apis.org/array-api/latest/search.html?q=as_xparray).

Reply:
The part around `copy, order` is from sklearn, `check_finite` is to cover our needs.

The function itself is not supposed to be in the standard, it's a helper for us. Basically it's `xp.asarray` but with extra things that we need or rely on for now.

Reply:
OK thanks.

> Basically it's `xp.asarray` but with extra things that we need or rely on for now.

Do we have an idea what might change (re: "for now") in the future?

I'm not too keen on the name `as_xparray` TBH, as it mixes separation by `_` with lack of separation between words. Also, the `xp` is not self-explanatory, except perhaps where it's meant for 1:1 replacement of `np`. I recognize that this is what the array API requires (though I couldn't find what it stands for or a rationale[^1]), but do we _need_ to name our methods after it?

[^1]: I looked on the array-api docs, repo & conformance test suite. It seems to just appear [here](https://data-apis.org/array-api/latest/purpose_and_scope.html#how-to-adopt-this-api).

I apologise if this has been discussed/decided already, I'd be happy to read up on it (I looked for it in the issues you linked too). If not, I think there's room for a better name. Conservatively just `as_xp_array`, but also `asarray_xp`, or even `asarray_api` (which contracts both `np.asarray` as well as `array-api`). It's a pity that `scipy.asarray` is already taken...

Reply:
> Do we have an idea what might change (re: "for now") in the future?

I am really not sure here. Ideally we would want to remove as much specification as possible to get closer to the standard Array API. This way we would not need this anymore as our code would just be compliant by default.

> I'm not too keen on the name as_xparray

Happy to revisit the name 😃 I don't have a strong preference on my side. I kind of like your proposal of `asarray_xp`.

Reply:
TBH it feels natural to fold this function into `xp.asarray`. I assume there was a prior discussion somewhere so that `array_api.asarray` does not get these extra keywords?